akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -313,6 +316,89 @@ private[stream] class VirtualPublisher[T]() extends Publisher[T] {
  */
+private[stream] case class MaterializedValueSource[M](
+  shape: SourceShape[M] = SourceShape[M](new Outlet[M]("Materialized.out")),
+  attributes: OperationAttributes = OperationAttributes.name("Materialized")) extends StreamLayout.Module {
+
+  override def subModules: Set[Module] = Set.empty
+  override def withAttributes(attr: OperationAttributes): Module = this.copy(shape = amendShape(attr), attributes = attr)
+  override def carbonCopy: Module = this.copy(shape = SourceShape(new Outlet[M]("Materialized.out")))
+
+  override def replaceShape(s: Shape): Module =
+    if (s == shape) this
+    else throw new UnsupportedOperationException("cannot replace the shape of MaterializedValueSource")
+
+  def amendShape(attr: OperationAttributes): SourceShape[M] = {
+    attr.nameOption match {
+      case None ⇒ shape
+      case s: Some[String] if s == attributes.nameOption ⇒ shape
+      case Some(name) ⇒ shape.copy(outlet = new Outlet(name + ".out"))
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] object MaterializedValuePublisher {
+  val NotRequested = 0
+  val Requested = 1
+  val Completed = 2
+}
+
+/**
+ * INTERNAL API
+ */
+private[stream] class MaterializedValuePublisher extends Publisher[Any] {
+  import MaterializedValuePublisher._
+
+  private val value = new AtomicReference[AnyRef](null)
+  private val registeredSubscriber = new AtomicReference[Subscriber[_ >: Any]](null)
+  private val requestState = new AtomicInteger(NotRequested)
+
+  def setValue(m: Any): Unit =
+    if (value.compareAndSet(null, m.asInstanceOf[AnyRef]) && requestState.get() == Requested)
+      pushAndClose(m)
+
+  private def pushAndClose(m: Any): Unit = {
+    if (requestState.compareAndSet(Requested, Completed)) {
+      val sub = registeredSubscriber.get()
+      ReactiveStreamsCompliance.tryOnNext(sub, m)
+      ReactiveStreamsCompliance.tryOnComplete(sub)
+    }
+  }
+
+  override def subscribe(subscriber: Subscriber[_ >: Any]): Unit = {
+    ReactiveStreamsCompliance.requireNonNullSubscriber(subscriber)
+    if (registeredSubscriber.compareAndSet(null, subscriber)) {
+      ReactiveStreamsCompliance.tryOnSubscribe(subscriber, new Subscription {
+        override def cancel(): Unit = requestState.set(Completed)
+        override def request(n: Long): Unit = {
+          if (n <= 0) {
+            ReactiveStreamsCompliance.tryOnError(
+              subscriber,
+              ReactiveStreamsCompliance.numberOfElementsInRequestMustBePositiveException)
+          } else {
+            if (requestState.compareAndSet(NotRequested, Requested)) {
+              val m = value.get()
+              if (m ne null) pushAndClose(m)
+            }
+          }
+        }
+      })
+    } else {
+      if (subscriber == registeredSubscriber.get())
+        ReactiveStreamsCompliance.rejectDuplicateSubscriber(subscriber)
+      else
+        ReactiveStreamsCompliance.rejectAdditionalSubscriber(subscriber, "MaterializedValuePublisher")
+    }
+  }
+}
```

Comment:
@drewhk just add this class in `akka-stream-tck-experimental` https://gist.github.com/ktoso/e91f71f7865e397c2d64
Can't test much but better than not being covered :)